tests/system_tests_dynamic_link_routes.py

```diff
@@ -0,0 +1,349 @@
+                                    {symbol('qd.link-route-patterns'):[u'linkRoute.prefix.#', symbol('in'),
+                                                                       u'some.*.pattern.#.end', symbol('out'),
+                                                                       u'another.pattern.*.x', symbol('inout')]},
+                                    [u'ClinkRoute.prefix', u'Fsome.*.pattern.#.end',
+                                     u'Eanother.pattern.*.x', u'Fanother.pattern.*.x'])
+        test.run()
+        self.assertEqual(None, test.error)
+
+
+    def test_11_prefix_with_multi_tenancy(self):
+        test = DynamicLinkRouteTest(self.routers[0].addresses[1],
+                                    self.routers[0].addresses[0],
+                                    self.routers[1].addresses[0],
+                                    {symbol('qd.link-route-patterns'):[u'linkRoute.prefix.#', symbol('inout')]},
+                                    [u'C0.0.0.0/linkRoute.prefix', u'D0.0.0.0/linkRoute.prefix'])
+        test.run()
+        self.assertEqual(None, test.error)
+
+
+    def test_12_pattern_with_multi_tenancy(self):
+        test = DynamicLinkRouteTest(self.routers[0].addresses[1],
+                                    self.routers[0].addresses[0],
+                                    self.routers[1].addresses[0],
+                                    {symbol('qd.link-route-patterns'):[u'linkRoute.#.prefix.#', symbol('inout')]},
+                                    [u'E0.0.0.0/linkRoute.#.prefix.#', u'F0.0.0.0/linkRoute.#.prefix.#'])
+        test.run()
+        self.assertEqual(None, test.error)
+
+
+class Entity(object):
+    def __init__(self, status_code, status_description, attrs):
+        self.status_code        = status_code
+        self.status_description = status_description
+        self.attrs              = attrs
+
+    def __getattr__(self, key):
+        return self.attrs[key]
+
+
+class RouterProxy(object):
+    def __init__(self, reply_addr):
+        self.reply_addr = reply_addr
+
+    def response(self, msg):
+        ap = msg.properties
+        return Entity(ap['statusCode'], ap['statusDescription'], msg.body)
+
+    def read_address(self, name):
+        ap = {'operation': 'READ', 'type': 'org.apache.qpid.dispatch.router.address', 'name': name}
+        return Message(properties=ap, reply_to=self.reply_addr)
+
+    def query_addresses(self):
+        ap = {'operation': 'QUERY', 'type': 'org.apache.qpid.dispatch.router.address'}
+        return Message(properties=ap, reply_to=self.reply_addr)
+
+
+class Timeout(object):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def on_timer_task(self, event):
+        self.parent.timeout()
+
+
+class PollTimeout(object):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def on_timer_task(self, event):
+        self.parent.poll_timeout()
+
+
+class DynamicLinkRouteTest(MessagingHandler):
```

Comment:
Can we please add a dynamic link route test which makes an outgoing connection? Connection properties containing link route information could be sent by a container that is accepting this connection initiated by the router.